subsys/net/ip/dhcpv6.c

```diff
@@ -0,0 +1,2205 @@
+		if (next_timeout < timeout_update) {
+			timeout_update = next_timeout;
+		}
+	}
+
+	k_mutex_unlock(&lock);
+
+	if (timeout_update != UINT64_MAX) {
+		if (now > timeout_update) {
+			timeout_update = 0ULL;
+		} else {
+			timeout_update -= now;
+		}
+
+		NET_DBG("Waiting for %llums", timeout_update);
+		k_work_reschedule(&dhcpv6_timeout_work, K_MSEC(timeout_update));
+	}
+}
+
+static void dhcpv6_iface_event_handler(struct net_mgmt_event_callback *cb,
+				       uint32_t mgmt_event, struct net_if *iface)
+{
+	sys_snode_t *node = NULL;
+
+	k_mutex_lock(&lock, K_FOREVER);
+
+	SYS_SLIST_FOR_EACH_NODE(&dhcpv6_ifaces, node) {
+		if (node == &iface->config.dhcpv6.node) {
+			break;
+		}
+	}
+
+	if (node == NULL) {
+		goto out;
+	}
+
+	if (mgmt_event == NET_EVENT_IF_DOWN) {
+		NET_DBG("Interface %p going down", iface);
+		dhcpv6_set_timeout(iface, UINT64_MAX);
+		dhcpv6_reschedule();
+	} else if (mgmt_event == NET_EVENT_IF_UP) {
+		NET_DBG("Interface %p coming up", iface);
+		dhcpv6_enter_state(iface, NET_DHCPV6_INIT);
+		dhcpv6_reschedule();
+	}
+
+out:
+	k_mutex_unlock(&lock);
+}
+
+static void dhcpv6_generate_client_duid(struct net_if *iface)
+{
+	struct net_linkaddr *lladdr = net_if_get_link_addr(iface);
+	struct net_dhcpv6_duid_storage *clientid = &iface->config.dhcpv6.clientid;
+	struct dhcpv6_duid_ll *duid_ll =
+				(struct dhcpv6_duid_ll *)&clientid->duid.buf;
+
+	memset(clientid, 0, sizeof(*clientid));
+
+	UNALIGNED_PUT(htons(DHCPV6_DUID_TYPE_LL), &clientid->duid.type);
+	UNALIGNED_PUT(htons(DHCPV6_HARDWARE_ETHERNET_TYPE), &duid_ll->hw_type);
+	memcpy(duid_ll->ll_addr, lladdr->addr, lladdr->len);
+
+	clientid->length = DHCPV6_DUID_LL_HEADER_SIZE + lladdr->len;
+}
+
+/* DHCPv6 public API */
+
+void net_dhcpv6_start(struct net_if *iface, struct net_dhcpv6_params *params)
+{
+	k_mutex_lock(&lock, K_FOREVER);
+
+	if (iface->config.dhcpv6.state != NET_DHCPV6_DISABLED) {
+		NET_ERR("DHCPv6 already running on iface %p, state %s", iface,
+			net_dhcpv6_state_name(iface->config.dhcpv6.state));
+		goto out;
+	}
+
+	if (!params->request_addr && !params->request_addr) {
+		NET_ERR("Information Request not supported yet");
+		goto out;
+	}
+
+	NET_DBG("Starting DHCPv6 on iface %p", iface);
+
+	iface->config.dhcpv6.params = *params;
+
+	if (sys_slist_is_empty(&dhcpv6_ifaces)) {
+		net_mgmt_add_event_callback(&dhcpv6_mgmt_cb);
+	}
+
+	sys_slist_append(&dhcpv6_ifaces, &iface->config.dhcpv6.node);
+
+	if (params->request_addr) {
+		iface->config.dhcpv6.addr_iaid = net_if_get_by_iface(iface);
+	}
+
+	if (params->request_prefix) {
+		iface->config.dhcpv6.prefix_iaid = net_if_get_by_iface(iface);
+	}
+
+	dhcpv6_generate_client_duid(iface);
+	dhcpv6_enter_state(iface, NET_DHCPV6_INIT);
+	dhcpv6_reschedule();
+
+out:
+	k_mutex_unlock(&lock);
+}
+
+void net_dhcpv6_stop(struct net_if *iface)
+{
+	k_mutex_lock(&lock, K_FOREVER);
+
+	switch (iface->config.dhcpv6.state) {
+	case NET_DHCPV6_DISABLED:
+		NET_ERR("DHCPv6 already disabled on iface %p", iface);
```

Comment:
Hmm, is this really an error? From application side it is, but from dhcpv6 client point of view this is just a no-op. Would warn or even info be enough here?

Reply:
True, downgraded to NET_INFO level.